### PR TITLE
Actions: Add change note checker

### PR DESCRIPTION
**Pull request by tausbn**
Checks whether a change note is present on a PR (assuming it should be) and adds a label to indicate when its not.

Considers all `.ql` and `.qll` files relevant, unless:
- the files are in a subdirectory of `experimental`, or
- the PR has been labelled with the label "Minor change"

If any relevant files are found, and there is no corresponding change note, the label "Needs change note" is applied to the PR.

I'm thinking we may want to roll this out to just the Python team at first. I have tested it manually, but I can't guarantee that it works perfectly.

Open questions:
- Should this add a comment instead of a label?
- Should this fail the action and block merging?

cc: @RasmusWL @calumgrant @adityasharad

### CODE REVIEW

**Comment by aschackmull** [comment]
I think we had something similar somewhere that worked like so: Anything without a change note that looks like it might need a change note (based on changed file paths, I guess) cannot be merged without a manually added label "don't-need-change-note" or something like that.
I think that seems like a slightly better workflow?

**Comment by aschackmull** [comment]
So I guess what I'm saying is that, yes, I think this should block merging, and that the "Minor change" label ought to be renamed to something including "change note" to make it clear what it means to apply that label.

**Comment by tausbn** [comment]
> So I guess what I'm saying is that, yes, I think this should block merging, and that the "Minor change" label ought to be renamed to something including "change note" to make it clear what it means to apply that label.

I was planning on putting this in the description of the label. I think "Doesn't need change note" is too long for a label.

**Comment by aschackmull** [comment]
https://github.com/github/codeql-go/labels uses a "no-change-note-required" label.

**Comment by aschackmull** [comment]
Possibly a good idea to check with the Go team how exactly they're using these change-note-related labels?

**Comment by tausbn** [comment]
Note to self: we should also exclude the `test` directories. Those things usually don't need change notes.

**Comment by RasmusWL** [comment]
> Should this fail the action and block merging?

yes

> https://github.com/github/codeql-go/labels uses a "no-change-note-required" label.

I would like a more clear name for the label than "Minor change" as well, so I'm all for using the "no-change-note-required" as well :+1:

**Comment by tausbn** [comment]
Any thoughts on feedback for the user? Is adding a label sufficient, or would it be better for it to add a comment instead?
(The problem with comments is that it'll keep adding comments for every change until you either add the `no-change-note-required` label or add the change note.)

**Comment by aschackmull** [comment]
CI fails for lots of reasons, so as long as you can click on the failed check and see why, then I don't think the PR needs any sort of feedback.

**Comment by RasmusWL** [comment]
> Any thoughts on feedback for the user? Is adding a label sufficient, or would it be better for it to add a comment instead?
(The problem with comments is that it'll keep adding comments for every change until you either add the no-change-note-required label or add the change note.)

I agree with `@aschackmull`, if the CI check fails, that is my expected level of feedback for the user (like the QLdoc check)

**Comment by tausbn** [comment]
Alright, in that case one more thing needs to be decided: when should the check trigger?
Currently it'll trigger on the default set of events (`opened`, `reopened`, and `synchronize`), and it'll definitely also need to trigger on `labeled` and `unlabeled` to handle the `no-change-note-required` label.

However, I think if it is to fail the CI, then it would be bothersome to already trigger when the PR is created. Imagine I'm creating a PR early in the process, before I even know _what_ the change note should say. I think it would be bothersome for the CI to fail outright, since that'll mean extra effort in order to figure out whether the CI failure is due to test issues (which is very relevant at that stage) or due to a missing change note (which is less relevant).

In other words, I'm leaning towards only triggering when the PR is not in draft mode.

**Comment by aschackmull** [comment]
> In other words, I'm leaning towards only triggering when the PR is not in draft mode.

Sounds good to me.

File: .github/workflows/check-change-note.yml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      - name: Check if change note file is present
+        uses: dorny/paths-filter@v2
+        id: paths_filter
+        with:
+          filters: |
+            change_note:
+              - '**/change-notes/*.md'
+      - name: Get PR labels
+        id: pr-labels
+        uses: joerick/pr-labels-action@v1.0.6
```
Comment (aeisenberg) [inline .github/workflows/check-change-note.yml:22]: For external actions, we always peg to a SHA, not a branch or tag. It's safer that way.

Comment (tausbn) [reply .github/workflows/check-change-note.yml]: Done!

File: .github/workflows/check-change-note.yml [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+      - name: Check if change note file is present
+        uses: dorny/paths-filter@7c0f15b688b020e95e00f15c61299b022f08ca95 # v2.8.0
+        id: paths_filter
+        with:
+          filters: |
+            change_note:
+              - '**/change-notes/*.md'
+      - name: Get PR labels
+        id: pr-labels
+        uses: joerick/pr-labels-action@0a4cc4ee0ab557ec0b1ae1157fa6fa7f9f4c494b # v1.0.6
```
Comment (adityasharad) [inline .github/workflows/check-change-note.yml:22]: I think you can avoid this step entirely, because there is a `labels` object in the Actions `pull_request` event payload. Can you try dropping this step, and using `if: !contains(github.event.pull_request_target.labels, 'no-change-note-required')` as the condition of the step below?

Comment (tausbn) [reply .github/workflows/check-change-note.yml:22]: Oh, cool. I'll try that. I found conflicting information about whether `.labels` worked on the internet, and the above seemed to do the trick.

(And in that case, I might just go all the way and avoid external actions entirely. I was trying to embrace the `import left-pad` spirit when implementing this. :laughing:)

Comment (adityasharad) [reply .github/workflows/check-change-note.yml:22]: I am generally of that mindset too: you can do quite a lot with just `gh api` to call the GitHub REST API (and it's already installed on the hosted runners).
My trick for working out what fields are in the event payload is simply to run workflows in a test repo, set an env var to hold `github.event`, and print it at the start. That said, the docs have recently improved: https://docs.github.com/en/actions/reference/events-that-trigger-workflows and https://docs.github.com/en/developers/webhooks-and-events/webhook-events-and-payloads.

Comment (tausbn) [reply .github/workflows/check-change-note.yml:22]: Aha, got it! What I needed was `github.event.pull_request.labels.*.name`. Very nice.

File: .github/workflows/check-change-note.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
+on:
+  pull_request_target:
```
[the remark this answers is not in the document]
Comment (tausbn) [reply .github/workflows/check-change-note.yml:2]: Hmm... So I _am_ using _some_ data from the payload (to wit, the list of files contained therein), but I guess that might be okay.

[the remark this answers is not in the document]
Comment (adityasharad) [reply .github/workflows/check-change-note.yml:2]: I think that's ok, especially as it's coming back from the API rather than directly from the event payload.

**Review by adityasharad — APPROVED** [review]
